0000-benchmark-suite.md

```diff
@@ -0,0 +1,782 @@
+we can have a relatively high degree of confidence in our measurements.
+
+### Randomizing Dynamic Allocation
+
+We should write a "shuffling allocator" wrapper that gives (most) heap
+allocations' addresses a uniform distribution. This helps us avoid measurement
+bias related to the accidental interaction between allocation order and
+caches. The Stabilizer paper shows that this can be done by maintaining a 256
+element buffer for each size class. When an allocation is requested make the
+underlying `malloc` call, choose a random integer `0 <= i < 256`, swap
+`buffer[i]` with the result of the `malloc`, and return the previous
+`buffer[i]`. Freeing a heap allocation involves choosing a random integer `0 <=
+i < 256`, swapping `buffer[i]` with the pointer we were given to free, and then
+freeing the previous `buffer[i]`. Implementing this is straightforward and I
+have a prototype already.
+
+For Wasmtime's dynamically mapped JIT code, we should investigate adding a Cargo
+feature that randomizes which pages we map for JIT code as well as the order of
+JIT'd function bodies placed within those pages.
+
+### Randomizing Stack Layout
+
+We should investigate adding a Cargo feature to Cranelift that adds a random
+amount of extra, unused space to each function's stack frame. This padding would
+always be a multiple of 16 bytes to keep the stack aligned, and would be at most
+4096 bytes in size. This helps us avoid measurement bias related to incidental
+stack layout.
+
+### Randomizing Environment Variables
+
+The benchmark runner should define new random environment variables for each
+benchmark program execution. This helps us avoid measurement bias related to
+incidental address space layout effects from environment variables.
+
+### Randomizing Link Order
+
+We should investigate building a "shuffling linker". Similar to how the
+shuffling allocator adds randomness by interposing between `malloc` calls and
+the underlying allocator, this would add randomness to the order in which object
+files are passed to the linker by interposing between `rustc` and `ld`. Our
+shuffling linker will look at the CLI arguments passed to it from `rustc`,
+shuffle the order of object files, shuffle object files within `.rlib`s (which
+are `.a` files that also contain a metadata file), and finally invoke the
+underlying linker. Before running benchmarks, we will create a few different
+`wasmtime` binaries that we choose from when sampling different benchmark
+executions, each with a different, randomized object file link order. This will
+help us avoid measurement bias related to getting "lucky" or "unlucky" with a
+single code layout due to the order of object files passed to the linker.
+
+> Note that this approach is more coarsely grained than what Stabilizer
+> achieves. Stabilizer randomized code layout at the function granularity, while
+> this approach works at the object file granularity.
```

Comment:
I believe lld has an dedicated option to specify the section order. In any case a linker script should work. Every function is put in a different section by rustc, so randomizing section order should be enough.